src/ChannelledStream.php

```diff
@@ -0,0 +1,112 @@
+<?php
+
+namespace Amp\Sync;
+
+use Amp\ByteStream\ClosableStream;
+use Amp\ByteStream\ReadableStream;
+use Amp\ByteStream\WritableStream;
+use Amp\ByteStream\StreamException;
+use Amp\Cancellation;
+use Amp\Pipeline\ConcurrentIterator;
+use Amp\Pipeline\Pipeline;
+use Amp\Serialization\Serializer;
+
+/**
+ * An asynchronous channel for sending data between threads and processes.
+ *
+ * Supports full duplex read and write.
+ *
+ * @template TReceive
+ * @template TSend
+ * @template-implements Channel<TReceive, TSend>
+ */
+final class ChannelledStream implements Channel, ClosableStream
```

Comment:
This is a `Channel` rather than a `Stream`, so it should probably be named `StreamChannel`?